qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        the samples.
+        - Lastly, if no special rules apply, the two ``Signal``s are multiplied generically via
+        multiplication of the envelopes as functions.
+
+    Args:
+        sig1: First signal.
+        sig2: Second signal.
+
+    Returns:
+        SignalSum: Representing the RHS of the formula when two Signals are multiplied.
+    """
+
+    # ensure signals are ordered from most to least specialized
+    sig1, sig2 = sort_signals(sig1, sig2)
+
+    if type(sig1) is Constant and type(sig2) is Constant:
+        return Constant(sig1(0.0) * sig2(0.0))
+    elif type(sig1) is Constant and type(sig2) is DiscreteSignal:
+        # multiply the samples by the constant
+        return DiscreteSignal(
+            dt=sig2.dt,
+            samples=sig1(0.0) * sig2.samples,
+            start_time=sig2.start_time,
+            duration=sig2.duration,
+            carrier_freq=sig2.carrier_freq,
+            phase=sig2.phase,
         )
+    elif type(sig1) is Constant and type(sig2) is Signal:
+        const = sig1(0.0)
+
+        def new_env(t):
+            return const * sig2.envelope(t)
+
+        return Signal(envelope=new_env, carrier_freq=sig2.carrier_freq, phase=sig2.phase)
+    elif type(sig1) is DiscreteSignal and type(sig2) is DiscreteSignal:
+        # verify compatible parameters before applying special rule
+        if (
+            sig1.start_time == sig2.start_time
+            and sig1.dt == sig2.dt
+            and len(sig1.samples) == len(sig2.samples)
+        ):
+            pwc1 = DiscreteSignal(
+                dt=sig2.dt,
+                samples=0.5 * sig1.samples * sig2.samples,
+                start_time=sig2.start_time,
+                duration=sig2.duration,
+                carrier_freq=sig1.carrier_freq + sig2.carrier_freq,
+                phase=sig1.phase + sig2.phase,
+            )
+            pwc2 = DiscreteSignal(
+                dt=sig2.dt,
+                samples=0.5 * sig1.samples * np.conjugate(sig2.samples),
+                start_time=sig2.start_time,
+                duration=sig2.duration,
+                carrier_freq=sig1.carrier_freq - sig2.carrier_freq,
+                phase=sig1.phase - sig2.phase,
+            )
+            return pwc1 + pwc2
+
+    # if no special cases apply, implement generic rule
+    def new_env1(t):
+        return 0.5 * sig1.envelope(t) * sig2.envelope(t)
+
+    def new_env2(t):
+        return 0.5 * sig1.envelope(t) * np.conjugate(sig2.envelope(t))
+
+    prod1 = Signal(
+        envelope=new_env1,
+        carrier_freq=sig1.carrier_freq + sig2.carrier_freq,
+        phase=sig1.phase + sig2.phase,
+    )
+    prod2 = Signal(
+        envelope=new_env2,
+        carrier_freq=sig1.carrier_freq - sig2.carrier_freq,
+        phase=sig1.phase - sig2.phase,
+    )
+    return prod1 + prod2
+
+
+def sort_signals(sig1: Signal, sig2: Signal) -> Tuple[Signal, Signal]:
```

Comment:
What does this enable?

Reply:
This is just to enable more succinct type checking for binary operations on signals that care about special cases. E.g. multiplication has a special rule for if one `sig1` and `sig2` is a `Constant` and the other is a more general `Signal`. By sorting `sig1` and `sig2` first, the multiplication routine only needs to check the single case: 
```
type(sig1) is Constant and type(sig2) is Signal
```